rpc_util_test.go

```diff
@@ -290,3 +293,138 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
 func BenchmarkGZIPCompressor1MiB(b *testing.B) {
 	bmCompressor(b, 1024*1024, NewGZIPCompressor())
 }
+
+// compressData compresses data using gzip and returns the compressed bytes.
+func compressData(data []byte) []byte {
+	var buf bytes.Buffer
+	gz := gzip.NewWriter(&buf)
+	_, _ = gz.Write(data)
+	_ = gz.Close()
+	return buf.Bytes()
+}
+
+// TestDecompress tests the decompress function with various scenarios, including
+// successful decompression, error handling, and edge cases like overflow or
+// premature data end. It ensures that the function behaves correctly with different
+// inputs, buffer sizes, and error conditions, using the "gzip" compressor for testing.
+func TestDecompress(t *testing.T) {
+	c := encoding.GetCompressor("gzip")
+
+	tests := []struct {
+		name                  string
+		compressor            encoding.Compressor
+		input                 mem.BufferSlice
+		maxReceiveMessageSize int
+		want                  mem.BufferSlice
+		compressedSize        int
+		error                 error
+	}{
+		{
+			name:       "Decompresses successfully with sufficient buffer size",
+			compressor: c,
+			input: func() mem.BufferSlice {
+				compressedData := compressData([]byte("decompressed data"))
+				return mem.BufferSlice{mem.NewBuffer(&compressedData, nil)}
+			}(),
+			maxReceiveMessageSize: 100,
+			want: func() mem.BufferSlice {
+				decompressed := []byte("decompressed data")
+				return mem.BufferSlice{mem.NewBuffer(&decompressed, nil)}
+			}(),
+			compressedSize: len([]byte("decompressed data")),
+			error:          nil,
+		},
+		{
+			name:                  "Fails with decompression error on empty input",
+			compressor:            c,
+			input:                 mem.BufferSlice{},
+			maxReceiveMessageSize: 1,
+			want:                  nil,
+			compressedSize:        0,
+			error:                 errors.New("decompression error"),
+		},
+		{
+			name:       "Fails with overflow error when decompressed size exceeds maxReceiveMessageSize",
+			compressor: c,
+			input: func() mem.BufferSlice {
+				compressedData := compressData([]byte("overflow test data"))
+				return mem.BufferSlice{mem.NewBuffer(&compressedData, nil)}
+			}(),
+			maxReceiveMessageSize: 5,
+			want:                  nil,
+			compressedSize:        6, //len([]byte("overflow test data")),
+			error:                 errors.New("overflow: received message size is larger than the allowed maxReceiveMessageSize (5 bytes)."),
+		},
+		{
+			name:       "Returns EOF error when maxReceiveMessageSize is MaxInt64 but data ends prematurely",
```

Comment:
EOF is not really failure. This test case is similar to case 1) except maxReceiveMessageSize is math.MAXInt64